File: src/ECSalinity.cpp
// Copyright (c) 2018 Justin Decker

//
// MIT License
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
//
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

#include "ECSalinity.h"

EC_Salinity::EC_Salinity(uint8_t i2c_address)
{
  _address = i2c_address;
#ifndef ARDUINO_SAMD_ZERO
  Wire.begin();
#endif // ifndef ARDUINO_SAMD_ZERO
}

EC_Salinity::EC_Salinity()
{
  _address = EC_SALINITY;
#ifndef ARDUINO_SAMD_ZERO
  Wire.begin();
#endif // ifndef ARDUINO_SAMD_ZERO
}

#ifdef ESP32
EC_Salinity::EC_Salinity(uint8_t sda, uint8_t scl, uint8_t i2c_address)
{
  _address = i2c_address;
  Wire.begin(sda, scl, 100000);
}

EC_Salinity::EC_Salinity(uint8_t sda, uint8_t scl)
{
  _address = EC_SALINITY;
  Wire.begin(sda, scl, 100000);
}

#endif // ifndef ESP32

// measurements
float EC_Salinity::_measure(bool EC, bool newTemp)
{
  if (newTemp)
  {
    measureTemp();
  }

  if (EC)
  {
    _send_command(EC_MEASURE_EC);
  }
  else
  {
    _send_command(EC_MEASURE_SW);
  }

  delay(EC_EC_MEASUREMENT_TIME);
  mS  = _read_register(EC_MS_REGISTER);
  raw = _read_register(EC_RAW_REGISTER);

  if (raw == 0.0)
  {
    mS = NAN; // make it NaN so the following statement will -1 everything
  }

  if (mS == mS)
  {
    PPM_500 = mS * 500;
    PPM_640 = mS * 640;
    PPM_700 = mS * 700;
    uS      = mS * 1000;
    S       = mS / 1000;

    salinityPSU = _read_register(EC_SALINITY_PSU);
  }
  else
  {
    mS          = -1;
    PPM_500     = -1;
    PPM_640     = -1;
    PPM_700     = -1;
    uS          = -1;
    S           = -1;
    salinityPSU = -1;
  }

  return mS;
}

float EC_Salinity::measureEC()
{
  return _measure(true, usingTemperatureCompensation());
}

float EC_Salinity::measureEC(bool newTemp)
{
  return _measure(true, newTemp);
}

float EC_Salinity::measureSW()
{
  _measure(false, usingTemperatureCompensation());
  return salinityPSU;
}

float EC_Salinity::measureSW(bool newTemp)
{
  _measure(false, newTemp);
  return salinityPSU;
}

float EC_Salinity::measureTemp()
{
  _send_command(EC_MEASURE_TEMP);
  delay(EC_TEMP_MEASURE_TIME);
  tempC = _read_register(EC_TEMP_REGISTER);
  if (tempC == -127.0)
  {
    tempF = -127;
  }
  else
  {
    tempF = ((tempC * 9) / 5) + 32;
  }
  return tempC;
}

// calibration
void EC_Salinity::calibrateEC(float solutionEC)
{
  _write_register(EC_SOLUTION_REGISTER, solutionEC);
  _send_command(EC_CALIBRATE_EC);
  delay(EC_EC_MEASUREMENT_TIME);
}

float EC_Salinity::getCalibrationEC()
{
  return _read_register(EC_CALIBRATE_EC_REGISTER);
}

void EC_Salinity::calibrateSW(float solutionSW)
{
  _write_register(EC_SOLUTION_REGISTER, solutionSW);
  _send_command(EC_CALIBRATE_SW);
  delay(EC_EC_MEASUREMENT_TIME);
}

float EC_Salinity::getCalibrationSW()
{
  return _read_register(EC_CALIBRATE_SW_REGISTER);
}

// temperature
void EC_Salinity::setTemp(float temp_C)
{
  _write_register(EC_TEMP_REGISTER, temp_C);
  tempC = temp_C;
  tempF = ((tempC * 9) / 5) + 32;
}

void EC_Salinity::setTempConstant(float b)
{
  _write_byte(EC_TEMP_COMPENSATION_REGISTER, b);
}

float EC_Salinity::getTempConstant()
{
  return _read_register(EC_TEMP_COMPENSATION_REGISTER);
}

void EC_Salinity::useTemperatureCompensation(bool b)
{
  uint8_t retval;
  uint8_t config = _read_byte(EC_CONFIG_REGISTER);

  if (b)
  {
    retval = bitSet(config, EC_TEMP_COMPENSATION_CONFIG_BIT);
  }
  else
  {
    retval = bitClear(config, EC_TEMP_COMPENSATION_CONFIG_BIT);
  }

  _write_byte(EC_CONFIG_REGISTER, retval);
}

bool EC_Salinity::usingTemperatureCompensation()
{
  uint8_t retval;

  retval = _read_byte(EC_CONFIG_REGISTER);
  return (retval >> EC_TEMP_COMPENSATION_CONFIG_BIT) & 0x01;
}

// utilities
uint8_t EC_Salinity::getVersion()
{
  return _read_byte(EC_VERSION_REGISTER);
}

uint8_t EC_Salinity::getFirmware()
{
  return _read_byte(EC_FW_VERSION_REGISTER);
}

void EC_Salinity::reset()
{
  _write_register(EC_CALIBRATE_EC_REGISTER, NAN);
  delay(10);
  _write_register(EC_CALIBRATE_SW_REGISTER, NAN);
  delay(10);
  setTempConstant(25);
  delay(10);
  useTemperatureCompensation(false);
}

void EC_Salinity::setI2CAddress(uint8_t i2cAddress)
{
  _write_register(EC_BUFFER_REGISTER, i2cAddress);
  _send_command(EC_I2C);
  _address = i2cAddress;
}

bool EC_Salinity::connected()
{
  uint8_t retval;

  retval = _read_byte(EC_VERSION_REGISTER);
  if (retval != 0xFF)
  {
    return true;
  }
  else
  {
    return false;
  }
}

float EC_Salinity::readEEPROM(uint8_t address)
{
  _write_register(EC_SOLUTION_REGISTER, address);
  _send_command(EC_READ);
  return _read_register(EC_BUFFER_REGISTER);
}

void EC_Salinity::writeEEPROM(uint8_t address, float value)
{
  _write_register(EC_SOLUTION_REGISTER, address);
  _write_register(EC_BUFFER_REGISTER,   value);
  _send_command(EC_WRITE);
}

//////////////////////////////////////////////////

void EC_Salinity::_change_register(uint8_t r)
{
  Wire.beginTransmission(_address);
  Wire.write(r);
  Wire.endTransmission();
  delay(10);
}

void EC_Salinity::_send_command(uint8_t command)
{
  Wire.beginTransmission(_address);
  Wire.write(EC_TASK_REGISTER);
  Wire.write(command);
  Wire.endTransmission();
  delay(10);
}

void EC_Salinity::_write_register(uint8_t reg, float f)
{
  uint8_t b[5];
  float   f_val = f;

  b[0] = reg;
  b[1] = *((uint8_t *)&f_val);
  b[2] = *((uint8_t *)&f_val + 1);
  b[3] = *((uint8_t *)&f_val + 2);
  b[4] = *((uint8_t *)&f_val + 3);
  Wire.beginTransmission(_address);
  Wire.write(b, 5);
  Wire.endTransmission();
  delay(10);
}

float EC_Salinity::_read_register(uint8_t reg)
{
  float retval;

  _change_register(reg);
  Wire.requestFrom(_address, (uint8_t)1);
  *((uint8_t *)&retval) = Wire.read();
  Wire.requestFrom(_address, (uint8_t)1);
  *((uint8_t *)&retval + 1) = Wire.read();
  Wire.requestFrom(_address, (uint8_t)1);
  *((uint8_t *)&retval + 2) = Wire.read();
  Wire.requestFrom(_address, (uint8_t)1);
  *((uint8_t *)&retval + 3) = Wire.read();
  delay(10);
  return retval;
}

void EC_Salinity::_write_byte(uint8_t reg, uint8_t val)
{
  uint8_t b[5];

  b[0] = reg;
  b[1] = val;
  Wire.beginTransmission(_address);
  Wire.write(b, 2);
  Wire.endTransmission();
  delay(10);
}

uint8_t EC_Salinity::_read_byte(uint8_t reg)
{
  uint8_t retval;

  _change_register(reg);
  Wire.requestFrom(_address, (uint8_t)1);
  retval = Wire.read();
  delay(10);
  return retval;
}

File: javascript/ECSalinity.js
var C = {
  EC_SALINITY: 0x3c,
  EC_MEASURE_EC: 80,
  EC_MEASURE_TEMP: 40,
  EC_CALIBRATE_PROBE: 20,
  EC_CALIBRATE_LOW: 10,
  EC_CALIBRATE_HIGH: 8,
  EC_CALCULATE_K: 2,
  EC_I2C: 1,
  EC_DRY: 82,
  EC_VERSION_REGISTER: 0,
  EC_MS_REGISTER: 1,
  EC_TEMP_REGISTER: 5,
  EC_K_REGISTER: 9,
  EC_SOLUTION_REGISTER: 13,
  EC_TEMPCOEF_REGISTER: 17,
  EC_CALIBRATE_REFHIGH_REGISTER: 21,
  EC_CALIBRATE_REFLOW_REGISTER: 25,
  EC_CALIBRATE_READHIGH_REGISTER: 29,
  EC_CALIBRATE_READLOW_REGISTER: 33,
  EC_CALIBRATE_OFFSET_REGISTER: 37,
  EC_SALINITY_PSU: 41,
  EC_DRY_REGISTER: 45,
  EC_TEMP_COMPENSATION_REGISTER: 49,
  EC_CONFIG_REGISTER: 50,
  EC_TASK_REGISTER: 51,
  EC_EC_MEASUREMENT_TIME: 250,
  EC_TEMP_MEASURE_TIME: 750,
  EC_DUALPOINT_CONFIG_BIT: 0,
  EC_TEMP_COMPENSATION_CONFIG_BIT: 1,
  PSU_TO_PPT_CONVERSION: 1.004715
};

function ECSalinity(_i2c, _address) {
  this.i2c = _i2c;
  this.address = _address;
}

ECSalinity.prototype._change_register = function(reg) {
  this.i2c.writeTo(this.address, reg);
  //for (f = 0; f < 10000; f++) {}
}

ECSalinity.prototype._send_command = function(cmd) {
  this.i2c.writeTo(this.address, [C.EC_TASK_REGISTER, cmd]);
  //for (f = 0; f < 10000; f++) {}
}

ECSalinity.prototype._read_byte = function(reg) {
  this._change_register(reg);
  return this.i2c.readFrom(this.address, 1)[0];
}

ECSalinity.prototype._write_byte = function(reg, val) {
  this.i2c.writeTo(this.address, [reg, val]);
  //for (f = 0; f < 10000; f++) {}
}

ECSalinity.prototype._read_register = function(reg) {
  var data = new Uint8Array(4);

  this._change_register(reg);
  data[0] = this.i2c.readFrom(this.address, 1)[0];
  data[1] = this.i2c.readFrom(this.address, 1)[0];
  data[2] = this.i2c.readFrom(this.address, 1)[0];
  data[3] = this.i2c.readFrom(this.address, 1)[0];

  return new Float32Array(data.buffer)[0];
}

ECSalinity.prototype._write_register = function(reg, val) {
  var data = new Float32Array(1);
  data[0] = val;

  this.i2c.writeTo(this.address, [reg, data.buffer[0], data.buffer[1], data.buffer[2], data.buffer[3]]);
  for (f = 0; f < 3000; f++) {}
}

exports.connect = function(_i2c, _address) {
  return new ECSalinity(_i2c, _address);
}

ECSalinity.prototype.measureTemp = function(callback) {
  var tempC;
  var tempF;
  var self = this;

  this._send_command(C.EC_MEASURE_TEMP);
  var i = setInterval(function() {
    tempC = self._read_register(C.EC_TEMP_REGISTER);
    tempF = ((tempC * 9) / 5) + 32;
    callback({
      tempC: tempC,
      tempF: tempF
    });
    clearInterval(i);
  }, C.EC_TEMP_MEASURE_TIME);
};

ECSalinity.prototype.measureEC = function(tempCoefficient, callback) {
  var self = this;

  this._write_register(C.EC_TEMPCOEF_REGISTER, tempCoefficient);
  this._send_command(C.EC_MEASURE_EC);

  var i = setInterval(function() {
    var mS = self._read_register(C.EC_MS_REGISTER);
    var PSU = self._read_register(C.EC_SALINITY_PSU);
    callback({
      mS: mS,
      uS: mS * 1000,
      S: mS / 1000,
      PPM_500: mS * 500,
      PPM_640: mS * 640,
      PPM_700: mS * 700,
      salinityPSU: self._read_register(C.EC_SALINITY_PSU),
      salinityPPT: PSU * C.PSU_TO_PPT_CONVERSION,
      salinityPPM: self.salinityPPT * 1000 * C.PSU_TO_PPT_CONVERSION
    });
    clearInterval(i);
  }, (C.EC_EC_MEASUREMENT_TIME));
}

ECSalinity.prototype.measureSalinity = function(tempCoefficient, callback) {

}

ECSalinity.prototype.calculateK = function(solutionEC, tempCoefficient, callback) {
  var self = this;
  var dualpoint = this.usingDualPoint();

  this.useDualPoint(false);
  this._write_register(C.EC_TEMPCOEF_REGISTER, tempCoefficient);
  this._write_register(C.EC_SOLUTION_REGISTER, solutionEC);
  this._send_command(C.EC_CALCULATE_K);

  var i = setInterval(function() {
    var K = self._read_register(C.EC_K_REGISTER);
    callback({
      K: K
    });
    clearInterval(i);
  }, (C.EC_EC_MEASUREMENT_TIME));

  this.useDualPoint(dualpoint)
}

ECSalinity.prototype.calibrateProbeLow = function(solutionEC, tempCoef, callback) {
  var self = this;
  var dualpoint = this.usingDualPoint();

  this.useDualPoint(false);
  this._write_register(C.EC_TEMPCOEF_REGISTER, tempCoef);
  this._write_register(C.EC_SOLUTION_REGISTER, solutionEC);
  this._send_command(C.EC_CALIBRATE_LOW);

  var i = setInterval(function() {
    var read = self._read_register(C.EC_CALIBRATE_READLOW_REGISTER);
    var reference = self._read_register(C.EC_CALIBRATE_REFLOW_REGISTER);
    callback({
      read: read,
      reference: reference
    });
    clearInterval(i);
  }, (C.EC_EC_MEASUREMENT_TIME));

  this.useDualPoint(dualpoint)
}

ECSalinity.prototype.calibrateProbeHigh = function(solutionEC, tempCoef, callback) {
  var self = this;
  var dualpoint = this.usingDualPoint();

  this.useDualPoint(false);
  this._write_register(C.EC_TEMPCOEF_REGISTER, tempCoef);
  this._write_register(C.EC_SOLUTION_REGISTER, solutionEC);
  this._send_command(C.EC_CALIBRATE_HIGH);

  var i = setInterval(function() {
    var read = self._read_register(C.EC_CALIBRATE_READHIGH_REGISTER);
    var reference = self._read_register(C.EC_CALIBRATE_REFHIGH_REGISTER);
    callback({
      read: read,
      reference: reference
    });
    clearInterval(i);
  }, (C.EC_EC_MEASUREMENT_TIME));

  this.useDualPoint(dualpoint)
}

ECSalinity.prototype.setDualPointCalibration = function(refLow, refHigh, readLow, readHigh) {
  this._write_register(C.EC_CALIBRATE_REFLOW_REGISTER, refLow);
  this._write_register(C.EC_CALIBRATE_REFHIGH_REGISTER, refHigh);
  this._write_register(C.EC_CALIBRATE_READLOW_REGISTER, readLow);
  this._write_register(C.EC_CALIBRATE_READHIGH_REGISTER, readHigh);
};

ECSalinity.prototype.calibrateDry = function(callback) {
  var self = this;

  this._send_command(C.EC_DRY);

  var i = setInterval(function() {
    var dry = self._read_register(C.EC_DRY_REGISTER);
    callback({
      dry: dry
    });
    clearInterval(i);
  }, (C.EC_EC_MEASUREMENT_TIME));

  this.useDualPoint(dualpoint)
}

ECSalinity.prototype.getVersion = function() {
  return this._read_byte(C.EC_VERSION_REGISTER);
};

ECSalinity.prototype.setK = function(k) {
  this._write_register(C.EC_K_REGISTER, k);
};

ECSalinity.prototype.getK = function() {
  return (Math.round(this._read_register(C.EC_K_REGISTER) * 100.0) / 100.0);
};

ECSalinity.prototype.getCalibrateDry = function() {
  return (Math.round(this._read_register(C.EC_DRY_REGISTER) * 100.0) / 100.0);
};

ECSalinity.prototype.getCalibrateOffset = function() {
  return Math.round(this._read_register(C.EC_CALIBRATE_OFFSET_REGISTER) * 100.0) / 100.0;;
};

ECSalinity.prototype.getCalibrateHigh = function() {
  return Math.round(this._read_register(C.EC_CALIBRATE_REFHIGH_REGISTER) * 100.0) / 100.0;;
};

ECSalinity.prototype.getCalibrateLow = function() {
  return Math.round(this._read_register(C.EC_CALIBRATE_REFLOW_REGISTER) * 100.0) / 100.0;;
};

ECSalinity.prototype.getCalibrateHighReading = function() {
  return Math.round(this._read_register(C.EC_CALIBRATE_READHIGH_REGISTER) * 100.0) / 100.0;;
};

ECSalinity.prototype.getCalibrateLowReading = function() {
  return Math.round(this._read_register(C.EC_CALIBRATE_READLOW_REGISTER) * 100.0) / 100.0;;
};

ECSalinity.prototype.reset = function() {
  this._write_register(C.EC_K_REGISTER, NaN);
  this._write_register(C.EC_CALIBRATE_OFFSET_REGISTER, NaN);
  this._write_register(C.EC_CALIBRATE_REFHIGH_REGISTER, NaN);
  this._write_register(C.EC_CALIBRATE_REFLOW_REGISTER, NaN);
  this._write_register(C.EC_CALIBRATE_READHIGH_REGISTER, NaN);
  this._write_register(C.EC_CALIBRATE_READLOW_REGISTER, NaN);
  this._write_register(C.EC_DRY_REGISTER, NaN);
  this.setTempConstant(0);
  this.useDualPoint(false);
  this.useTemperatureCompensation(false);
};

ECSalinity.prototype.setTempConstant = function(temp) {
  this._write_byte(C.EC_TEMP_COMPENSATION_REGISTER, temp);
};

ECSalinity.prototype.setI2CAddress = function(i2cAddress) {
  this._write_register(C.EC_SOLUTION_REGISTER, i2cAddress);
  this._send_command(C.EC_I2C);
  this.address = i2cAddress;
};

ECSalinity.prototype.getTempConstant = function() {
  return this._read_byte(C.EC_TEMP_COMPENSATION_REGISTER);
};

ECSalinity.prototype.useTemperatureCompensation = function(b) {
  var retval = this._read_byte(C.EC_CONFIG_REGISTER);

  retval = this._bit_set(retval, C.EC_TEMP_COMPENSATION_CONFIG_BIT, b);
  this._write_byte(C.EC_CONFIG_REGISTER, retval);
};

ECSalinity.prototype.useDualPoint = function(b) {
  var retval = this._read_byte(C.EC_CONFIG_REGISTER);

  retval = this._bit_set(retval, C.EC_DUALPOINT_CONFIG_BIT, b);
  this._write_byte(C.EC_CONFIG_REGISTER, retval);
};

ECSalinity.prototype.usingTemperatureCompensation = function() {
  var retval = this._read_byte(C.EC_CONFIG_REGISTER);
  return (retval >> 1) & 0x01;
};

ECSalinity.prototype.usingDualPoint = function() {
  var retval = this._read_byte(C.EC_CONFIG_REGISTER);
  return (retval >> 0) & 0x01;
};

ECSalinity.prototype._bit_set = function(v, index, x) {
  var mask = 1 << index;
  v &= ~mask;
  if (x) {
    v |= mask;
  }
  return v;
};